src/main/kotlin/org/rust/lang/core/lexer/RustLexer.flex

```diff
@@ -120,18 +122,17 @@ CHAR_LITERAL   = ( \' ( [^\\\'\r\n] | \\[^\r\n] | "\\x" [a-fA-F0-9]+ | "\\u{" [a
                | ( \' [\p{xidcontinue}]* \' {SUFFIX}? )
 STRING_LITERAL = \" ( [^\\\"] | \\[^] )* ( \" {SUFFIX}? | \\ )?
 
-
-///////////////////////////////////////////////////////////////////////////////////////////////////
-// Other
-///////////////////////////////////////////////////////////////////////////////////////////////////
-
-SHEBANG_LINE=\#\![^\[].*
-
 %%
 
-<YYINITIAL> \'                    { yybegin(IN_LIFETIME_OR_CHAR); yypushback(1); }
+<YYINITIAL>                     {
+  "#!" [\r\n]                     { yybegin(INITIAL); yypushback(1); return SHEBANG_LINE; }
+  "#!" [^\[\r\n] [^\r\n]*         { yybegin(INITIAL); return SHEBANG_LINE; }
+  [^]                             { yybegin(INITIAL); yypushback(1); }
```

Comment:
There is one extreme case when this lexer behaves differently than rustc ([playground](http://is.gd/hU30nI)):
```rust
#!<<EOF>>
```
Though, I don't think handling this specifically is worth anything.